### PR TITLE
Phase2 workflow fixes

### PR DESCRIPTION
**Pull request by kpedro88**
This is the promised "workflow fix" PR for Phase2. This is just meant to get some things nominally working before 8_1_0_pre3. Further reorganizations will come later.

Major changes:
1. local reco workflows now use full local reco instead of just tracker local reco
2. 2023 workflows now use the Phase2 era
3. rename 2023dev -> 2023tilted

Various bug fixes/customizations/patches:
1. turn off `useLumiInfoRunHeader` in `ecalMultiFitUncalibRecHit` to fix ECAL `bunchSpacingProducer` exception
2. turn off all "recover" options in `ecalRecHit` to avoid use of [EcalRecHitWorkerRecover](https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/RecoLocalCalo/EcalRecProducers/plugins/EcalRecHitWorkerRecover.cc), which has non-optional use of EcalEndcapGeometryRecord that causes an exception in Phase2 geometries (a better fix for this should come later)
3. additional ZDC customization to handle lack of DIGI2RAW/RAW2DIGI for HCAL
4. temporary fix from @jshlee to prevent ME0 crash

Known warnings:
1. Expected since HEback isn't set up in the hexagonal geometry yet:

```
%MSG-e HGCDigitizer:  MixingModule:mix 20-Apr-2016 02:57:06 CDT Run: 1 Event: 1
 @ accumulate : can't find HGCHitsHEback collection of g4SimHits
```

2. CSC people are looking into this one, the actual file address is `/cvmfs/cms-ib.cern.ch/2016-17/slc6_amd64_gcc530/cms/CSCTrackFinderEmulation/1.2/data/L1Trigger/CSCTrackFinder/data/core_2014_05_15/comp_dphi_8.dat` with a slash between data and L1Trigger:

```
Cannot read file: /cvmfs/cms-ib.cern.ch/2016-17/slc6_amd64_gcc530/cms/CSCTrackFinderEmulation/1.2/dataL1Trigger/CSCTrackFinder/data/core_2014_05_15/comp_dphi_8.dat, addr: 0
```

I tested this using the normal and tilted FourMuPt1_200 upgrade workflows:

```
runTheMatrix.py -w upgrade -l 11000.0 > & matrix_11000.log &
runTheMatrix.py -w upgrade -l 10600.0 > & matrix_10600.log &
```

10 events run to completion for steps 1 and 2 with no crashes or exceptions.

attn: @boudoul, @lgray, @bsunanda, @ianna, @calabria

@davidlange6 - should we add upgrade workflows into the PR/IB standard tests in this PR, or wait?

### CODE REVIEW

**Comment by kpedro88** [comment]
@cmsbuild please test

File: RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py [hunk 1]
```diff
@@ -63,3 +63,6 @@
       chi2ThreshEE_ = cms.double(50.0),
    )                                           
 )
+
+from Configuration.StandardSequences.Eras import eras
+eras.phase2_common.toModify( ecalMultiFitUncalibRecHit.algoPSet, useLumiInfoRunHeader = cms.bool(False) )
```
Comment (slava77) [inline RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py]: why not?

File: RecoLocalMuon/GEMSegment/plugins/ME0SegFit.cc [hunk 2]
```diff
@@ -343,7 +343,8 @@ ME0SegFit::SMatrixSym12 ME0SegFit::weightMatrix() {
   int row = 0;
   
   for (ME0SetOfHits::const_iterator it = hits_.begin(); it != hits_.end(); ++it) {
-    
+    if (row > 11) break; // temp due to max size of matrix
```
Comment (slava77) [inline RecoLocalMuon/GEMSegment/plugins/ME0SegFit.cc]: it looks like there is a bug in the code, if it comes to this point.
Maybe better not to run this?

File: RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py (hunk 1 above)
Comment (kpedro88) [inline RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py]: The underlying problem is that bunchSpacingProducer is separate from localreco:
https://github.com/cms-sw/cmssw/blob/CMSSW_8_1_X/Configuration/StandardSequences/python/Reconstruction_cff.py#L101

Without it, there is an exception when running ecalMultiFitUncalibRecHit. The above customization seemed like the easiest way to prevent the exception. I'm open to other solutions.

File: RecoLocalMuon/GEMSegment/plugins/ME0SegFit.cc (hunk 2 above)
Comment (kpedro88) [inline RecoLocalMuon/GEMSegment/plugins/ME0SegFit.cc]: This is a temporary fix from @jshlee to avoid a segmentation violation. He or another muon developer will submit a better fix in a subsequent PR.

File: RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py (hunk 1 above)
Comment (slava77) [inline RecoLocalCalo/EcalRecProducers/python/ecalMultiFitUncalibRecHit_cfi.py]: add it to the sequence you are running (at the start).
It's already in the localreco in HI and cosmics.
There is no harm from listing the same module several times in different sequences

**Comment by boudoul** [comment]
looks like the errors are not related to this PR (other PRs  are suffering from the same errors ..)

**Comment by kpedro88** [comment]
@boudoul - yes, I've seen them in other PRs also. It appears to be a problem with jet corrections:
```
An exception of category 'NoRecord' occurred while
   [0] Processing run: 165121 lumi: 62 event: 23609118
   [1] Running path 'reconstruction_step'
   [2] Calling event method for module L1FastjetCorrectorProducer/'ak4PFL1FastjetCorrector'
Exception Message:
No "JetCorrectionsRecord" record found in the EventSetup for synchronization value
Run: 165121 LuminosityBlock: 62 Event: 0 Time: 5607255048184066530
 Please add an ESSource or ESProducer that delivers such a record.
```

**Comment by hengne** [comment]
@boudoul @kpedro88 it is xrootd access to the input files.. my PR is also suffering this since yesterday evening: https://github.com/cms-sw/cmssw/pull/14098

**Comment by kpedro88** [comment]
@slava77 latest commit addresses your recommendation about bunchSpacingProducer

File: Configuration/StandardSequences/python/Reconstruction_cff.py [hunk 3]
```diff
@@ -194,4 +196,5 @@
 #
 reconstruction_standard_candle = cms.Sequence(localreco*globalreco*vertexreco*recoJetAssociations*btagging*electronSequence*photonSequence)
 
-
+from Configuration.StandardSequences.Eras import eras
+eras.phase2_common.toReplaceWith( localreco, _phase2_localreco )
```
Comment (slava77) [inline Configuration/StandardSequences/python/Reconstruction_cff.py]: maybe easier to add globally, no eras

Comment (kpedro88) [inline Configuration/StandardSequences/python/Reconstruction_cff.py]: is that allowed? I don't want to screw up other workflows

Comment (slava77) [inline Configuration/StandardSequences/python/Reconstruction_cff.py]: should be OK, we'll see in the tests anyways.

**Comment by kpedro88** [comment]
Update on known warning 2: it's unrelated to Phase2, caused by a change in the external. I submitted an issue against the external: https://github.com/cms-externals/CSCTrackFinderEmulation/issues/3

File: Configuration/StandardSequences/python/Reconstruction_cff.py (hunk 3 above)
Comment (kpedro88) [inline Configuration/StandardSequences/python/Reconstruction_cff.py]: submitted as a separate commit so it can be reverted back to the Era-dependent version if necessary

**Comment by smuzaffar** [comment]
@kpedro88 , cms-externals/CSCTrackFinderEmulation tool file is updated to have the correct path. Next IB (23h00) should have the fix.

**Comment by kpedro88** [comment]
@smuzaffar great, thanks

**Comment by kpedro88** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
The comparison looks good to me.

@davidlange6 @slava77 @civanch @ianna @hengne - can you sign ASAP so we can make 810pre3?

**Comment by davidlange6** [comment]
all - will merge this tonight as to build pre3 unless there are issues raised in the meantime

**Comment by kpedro88** [comment]
@davidlange6 did you mean to merge this?

**Comment by kpedro88** [comment]
Ah I see, it was included in #14181
